crates/wasm-encoder/src/reencode.rs

```diff
@@ -0,0 +1,1751 @@
+    }
+
+    fn field_type(
+        &mut self,
+        field_ty: wasmparser::FieldType,
+    ) -> Result<crate::FieldType, Error<Self::Error>> {
+        utils::field_type(self, field_ty)
+    }
+
+    fn func_type(
+        &mut self,
+        func_ty: wasmparser::FuncType,
+    ) -> Result<crate::FuncType, Error<Self::Error>> {
+        utils::func_type(self, func_ty)
+    }
+
+    fn global_type(
+        &mut self,
+        global_ty: wasmparser::GlobalType,
+    ) -> Result<crate::GlobalType, Error<Self::Error>> {
+        utils::global_type(self, global_ty)
+    }
+
+    fn heap_type(
+        &mut self,
+        heap_type: wasmparser::HeapType,
+    ) -> Result<crate::HeapType, Error<Self::Error>> {
+        utils::heap_type(self, heap_type)
+    }
+
+    fn instruction<'a>(
+        &mut self,
+        arg: wasmparser::Operator<'a>,
+    ) -> Result<crate::Instruction<'a>, Error<Self::Error>> {
+        utils::instruction(self, arg)
+    }
+
+    fn memory_type(&mut self, memory_ty: wasmparser::MemoryType) -> crate::MemoryType {
+        utils::memory_type(self, memory_ty)
+    }
+
+    fn mem_arg(&mut self, arg: wasmparser::MemArg) -> crate::MemArg {
+        utils::mem_arg(self, arg)
+    }
+
+    fn ordering(&mut self, arg: wasmparser::Ordering) -> crate::Ordering {
+        utils::ordering(self, arg)
+    }
+
+    fn ref_type(
+        &mut self,
+        ref_type: wasmparser::RefType,
+    ) -> Result<crate::RefType, Error<Self::Error>> {
+        utils::ref_type(self, ref_type)
+    }
+
+    fn storage_type(
+        &mut self,
+        storage_ty: wasmparser::StorageType,
+    ) -> Result<crate::StorageType, Error<Self::Error>> {
+        utils::storage_type(self, storage_ty)
+    }
+
+    fn struct_type(
+        &mut self,
+        struct_ty: wasmparser::StructType,
+    ) -> Result<crate::StructType, Error<Self::Error>> {
+        utils::struct_type(self, struct_ty)
+    }
+
+    fn sub_type(
+        &mut self,
+        sub_ty: wasmparser::SubType,
+    ) -> Result<crate::SubType, Error<Self::Error>> {
+        utils::sub_type(self, sub_ty)
+    }
+
+    fn table_type(
+        &mut self,
+        table_ty: wasmparser::TableType,
+    ) -> Result<crate::TableType, Error<Self::Error>> {
+        utils::table_type(self, table_ty)
+    }
+
+    fn tag_kind(&mut self, kind: wasmparser::TagKind) -> crate::TagKind {
+        utils::tag_kind(self, kind)
+    }
+
+    fn tag_type(&mut self, tag_ty: wasmparser::TagType) -> crate::TagType {
+        utils::tag_type(self, tag_ty)
+    }
+
+    fn val_type(
+        &mut self,
+        val_ty: wasmparser::ValType,
+    ) -> Result<crate::ValType, Error<Self::Error>> {
+        utils::val_type(self, val_ty)
+    }
+
+    fn custom_section<'a>(
+        &mut self,
+        section: wasmparser::CustomSectionReader<'a>,
+    ) -> crate::CustomSection<'a> {
+        utils::custom_section(self, section)
+    }
+
+    fn intersperse_section_hook(
```

Comment:
Since this is somewhat subtle mind adding some docs? For example I thought `before` should be just `SectionId` and not `Option` but I see now that it's used to indicate the end of the module with `None`

Reply:
One aspect I'm not set on yet is whether custom sections should be reported as before and after, since one likely does not care whether one is between custom sections but rather wants to know which non-custom section was last and is next

Reply:
Yeah I think it's reasonable to document that one of after or before will be `None` or a non-custom section and not worry about between custom sections.

Reply:
Done :)